camera-viewfinder-compose/src/main/java/com/google/jetpackcamera/viewfinder/surface/SurfaceTransformationUtil.kt

```diff
@@ -0,0 +1,207 @@
+        )
+    }
+
+    private fun setMatrixRectToRect(
+        matrix: Matrix, source: RectF, destination: RectF,
+    ) {
+        val matrixScaleType = Matrix.ScaleToFit.CENTER
+        // android.graphics.Matrix doesn't support fill scale types. The workaround is
+        // mapping inversely from destination to source, then invert the matrix.
+        matrix.setRectToRect(destination, source, matrixScaleType)
+        matrix.invert(matrix)
+    }
+
+    private fun getPreviewViewViewportRectForMismatchedAspectRatios(
+        transformationInfo: SurfaceRequest.TransformationInfo,
+        previewViewSize: Size,
+        resolution: Size
+    ): RectF {
+        val previewViewRect = RectF(
+            0f, 0f, previewViewSize.width.toFloat(),
+            previewViewSize.height.toFloat()
+        )
+        val rotatedViewportSize = getRotatedViewportSize(transformationInfo, resolution)
+        val rotatedViewportRect = RectF(
+            0f, 0f, rotatedViewportSize.width.toFloat(),
+            rotatedViewportSize.height.toFloat()
+        )
+        val matrix = Matrix()
+        setMatrixRectToRect(
+            matrix,
+            rotatedViewportRect,
+            previewViewRect,
+        )
+        matrix.mapRect(rotatedViewportRect)
+        return rotatedViewportRect
+    }
+
+    @SuppressLint("RestrictedApi")
+    fun getSurfaceToPreviewViewMatrix(
+        previewViewSize: Size,
+        transformationInfo: SurfaceRequest.TransformationInfo,
+        isFrontCamera: Boolean,
+        resolution: Size
+    ): Matrix {
+        // Get the target of the mapping, the coordinates of the crop rect in PreviewView.
+        val previewViewCropRect: RectF =
+            if (isViewportAspectRatioMatchPreviewView(
+                    transformationInfo,
+                    previewViewSize,
+                    resolution
+                )
+            ) {
+                // If crop rect has the same aspect ratio as PreviewView, scale the crop rect to fill
+                // the entire PreviewView. This happens if the scale type is FILL_* AND a
+                // PreviewView-based viewport is used.
+                RectF(
+                    0f, 0f, previewViewSize.width.toFloat(),
+                    previewViewSize.height.toFloat()
+                )
+            } else {
+                // If the aspect ratios don't match, it could be 1) scale type is FIT_*, 2) the
+                // Viewport is not based on the PreviewView or 3) both.
+                getPreviewViewViewportRectForMismatchedAspectRatios(
+                    transformationInfo, previewViewSize, resolution
+                )
+            }
+        val resolutionRect = RectF(0F, 0F, resolution.width.toFloat(), resolution.height.toFloat())
+        val matrix = TransformUtils.getRectToRect(
+            resolutionRect, previewViewCropRect, transformationInfo.rotationDegrees
+        )
+        if (isFrontCamera && transformationInfo.hasCameraTransform()) {
```

Comment:
There seems to be an inssue when switching to "single stream" where the image gets stretched. I think it may have something to do with rotation somehow affecting the expected aspect ratio (i.e., 16:9 rather than 9:16). We may need to do a similar `if` condition to this one where we check `hasCameraTransform` to see if rotation has already been applied so we scale along the correct axes.